### DISCUSSION TITLE
Redirect to old page after login

### DISCUSSION THREAD

**Discussion by iammathew** [ory/kratos#2247] [Q&A] [answered]
Hey!

I'm currently trying to implement a custom UI for Ory Kratos, and running into the following issue, when a user gets redirected to login by accessing the settings page (for your information I'm running Next.js SSR as a UI in this example).

1. User opens `/settings` and Next.js notices missing flowId therefore redirecting to `/api/.ory/self-service/settings/browser`, which should create the flow and redirect back to the settings UI.
2. Ory Kratos notices the user did not yet authenticate, therefore redirects to the login UI `/login` instead of back to the settings UI
3. User logs in but ends up in the dashboard instead of the settings page

My quick analysis:
When Ory itself redirects to the login page (when trying to create another flow) it does not include a special return_to parameter, therefore it will use the default return URL.

Is there any way of sending back the user to the `/settings` or more correct I guess `/api/.ory/self-service/settings/browser` page here?

**Comment by aeneasr** [comment, marked as answer]
If the initial `/api/.ory/self-service/settings/browser` contains a `return_to` (so e.g. `/api/.ory/self-service/settings/browser?return_to=/settings` it *should* be propagated to the login initialization as well! If not please open an issue for this with a minimal reproducible example, or, even better, a test case in one of these files: https://github.com/ory/kratos/tree/master/test/e2e/cypress/integration/profiles/email/settings

**Reply by iammathew** [reply]
Thanks for the quick answer!

To be quite frank, I don't think that is the case :)

<img width="711" alt="Screen Shot 2022-02-22 at 16 16 13" src="https://user-images.githubusercontent.com/23657891/155163408-989453b6-139b-4834-82f5-8b55ed8ea500.png">

As you can see in the network inspection it just redirects to `/login` the specified default login UI inside the configuration.



Also if I understand the semantics of the `return_to` argument correctly this would also not be correct behavior.
`return_to` redirects the user after the requested flow is completed, which means the action is sent and successful.
If it would just propagate this query argument to the redirection to log in, it would redirect there after the login flow completed (but initially one intended to be redirected there after one changed the settings).

I will try to create a quick reproducible example!

Cheers

**Reply by MilesNorton** [reply]
Before trying my solution, I've looked at your network request and can see the `return_to=/settings`, my understanding is that it has to be a full address including http/https.

**Reply by aeneasr** [reply]
We've made some improvement to the `return_to` handling but it's ossible that this specific use case still doesn't work. Please feel free to create an issue if that is the case!

**Comment by MilesNorton** [comment]
@iammathew Hey mate, a way around this could be to route to `/settings` and not to `.ory/self-service/settings/browser` first.
So on the settings page you can first validate that they are authed yourself, and if they are not you can manually redirect them to the login page with the return_to param that was used for accessing the `/settings` route. If they are authed you can then request the flow from `.ory/self-service/settings/browser` and render the form. Happy to provide some code snippets in nextjs if you'd like
